prairielib/lib/migrations.js

```diff
@@ -34,136 +35,272 @@ module.exports.init = function (dir, proj, callback) {
+  let allMigrationsHaveTimestamp = true;
+  let seenIndexes = new Set();
+  let seenTimestamps = new Set();
+  for (const migration of migrations) {
+    const { filename, timestamp, index } = migration;
+
+    if (timestamp !== null) {
+      if (seenTimestamps.has(timestamp)) {
+        throw new Error(`Duplicate migration timestamp: ${timestamp} (${filename})`);
+      }
+      seenTimestamps.add(timestamp);
+      hasSeenTimestamp = true;
+    } else {
+      allMigrationsHaveTimestamp = false;
+    }
+
+    if (index !== null) {
+      if (seenIndexes.has(index)) {
+        throw new Error(`Duplicate migration index: ${index} (${filename})`);
+      }
+      seenIndexes.add(index);
+      hasSeenIndex = true;
+    } else {
+      allMigrationsHaveIndex = false;
+    }
+  }
+
+  // Validation: if one migration has a timestamp, *all* migrations must have a timestamp.
+  if (hasSeenTimestamp && !allMigrationsHaveTimestamp) {
+    const filesMissingTimestamp = migrations
+      .filter((m) => m.timestamp === null)
+      .map((m) => m.filename)
+      .join(', ');
+    throw new Error(
+      `The following migration files are missing timestamps: ${filesMissingTimestamp}`
+    );
+  }
+
+  // Validation: if one migration has an index, *all* migrations must have an index.
+  if (hasSeenIndex && !allMigrationsHaveIndex) {
+    const filesMissingIndex = migrations
+      .filter((m) => m.index === null)
+      .map((m) => m.filename)
+      .join(', ');
+    throw new Error(`The following migration files are missing indexes: ${filesMissingIndex}`);
+  }
+
+  return migrations;
+}
+
+/**
+ * Determine the ordering of migrations. We will have already validates that either all migrations
+ * have a timestamp or none of them do; ditto for indexes. So we can safely order on either
+ * one or the other. Default to the timestamp if they're available, otherwise order on the index.
+
+ * @param {MigrationFile[]} migrationFiles
+ * @return {MigrationFile[]}
+ */
+function sortMigrationFiles(migrationFiles) {
+  let allMigrationsHaveTimestamp = migrationFiles.every((m) => m.timestamp);
+  return migrationFiles.sort((a, b) => {
+    if (allMigrationsHaveTimestamp) {
+      return a.timestamp.localeCompare(b.timestamp);
+    } else {
+      return a.index - b.index;
+    }
+  });
+}
+
+/**
+ *
+ * @param {MigrationFile[]} migrationFiles
+ * @param {{ index: number | null, timestamp: string | null }[]} executedMigrations
+ * @return {MigrationFile[]}
+ */
+function getMigrationsToExecute(migrationFiles, executedMigrations) {
+  // If no migrations have ever been run, run them all.
+  if (executedMigrations.length === 0) {
+    return migrationFiles;
+  }
+
+  // If our migration files have timestamps, use them to determine which
+  // have or have not executed. Otherwise, use the index.
+  if (migrationFiles.every((m) => m.timestamp)) {
+    return migrationFiles.filter(
+      (m) => !executedMigrations.some((e) => e.timestamp === m.timestamp)
```

Comment:
Use a `Set` to avoid the O(N^2) filter? As in https://github.com/PrairieLearnInc/PrairieTest/commit/168c783e62aa9d2f36980610b3d73ff9506ab1da

Reply:
Fixed up in 43ea2d8.